src/main/scala/vsys/blockchain/state/opcdiffs/BasicOpcDiff.scala

```diff
@@ -4,109 +4,137 @@ import vsys.blockchain.transaction.ValidationError
+  def sqrtBigInt(y: BigInt): BigInt = {
+    if (y > 3) Stream.iterate((y, (y >> 1) + 1)){ case (z, x) => (x, (y / x + x) >> 1) }.dropWhile{ case(z, x) => x < z }.head._1
+    else if (y > 0) 1 else 0
+  }
+
+  // not converting byte to boolean
+  case class BoolByteBiOperator(val op: (Byte, Byte) => Byte)
+
+  val bAnd = BoolByteBiOperator(_ & _)
+  val bOr  = BoolByteBiOperator(_ | _)
+  val bXor = BoolByteBiOperator(_ ^ _)
+
+  val bNot: Byte => Byte = x => ~x & 1
+
+  private implicit def try2Either(t: Try[Either[ValidationError, DataEntry]]): Either[ValidationError, DataEntry] =
+    t.recover({ case _ => Left(OverflowError) }).get
+
+  def numBiOperation(x: DataEntry, y: DataEntry, operator: NumBiOperator): Either[ValidationError, DataEntry] =
+    if (x.dataType == y.dataType) x.dataType match {
+      case Int32      => operator.int(x, y).map(formatRes[Int](_, Int32))
+      case Amount     => operator.long(x, y).map(formatRes[Long](_, Amount))
+      case Timestamp  => operator.long(x, y).map(formatRes[Long](_, Timestamp))
+      case BigInteger => operator.bigInt(x, y).map(formatRes[BigInt](_, BigInteger))
       case _ => Left(ContractUnsupportedOPC)
     } else Left(ContractDataTypeMismatch)
 
-  def sqrt(x: DataEntry, dataStack: Seq[DataEntry], pointer: Byte): Either[ValidationError, Seq[DataEntry]] =
-    if (pointer > dataStack.length || pointer < 0) Left(ContractLocalVariableIndexOutOfRange)
-    else x.dataType match {
+  def sqrt(x: DataEntry): Either[ValidationError, DataEntry] =
+    x.dataType match {
       case BigInteger => {
         val value = BigInteger.deserializer(x.data)
         if (value < 0) Left(ValidationError.OverflowError)
-        else appendRes[BigInt](pointer, sqrtBigInt(value),  BigInteger, dataStack)
+        else formatRes[BigInt](sqrtBigInt(value), BigInteger)
       }
       case _ => Left(ContractUnsupportedOPC)
     }
 
-  // babylonian method
-  def sqrtBigInt(y: BigInt): BigInt = {
-    if (y > 3) Stream.iterate((y, (y >> 1) + 1)){ case (z, x) => (x, (y / x + x) >> 1) }.dropWhile{ case(z, x) => x < z }.head._1
-    else if (y > 0) 1 else 0
-  }
-  def concat(x: DataEntry, y: DataEntry, dataStack: Seq[DataEntry], pointer: Byte): Either[ValidationError, Seq[DataEntry]] = {
-    if (pointer > dataStack.length || pointer < 0) {
-      Left(ContractLocalVariableIndexOutOfRange)
-    } else {
-      for {
-        res <- DataEntry.create(x.data ++ y.data, DataType.ShortBytes)
-      } yield dataStack.patch(pointer, Seq(res), 1)
+  def boolBiOperation(x: DataEntry, y: DataEntry, operator: BoolByteBiOperator): Either[ValidationError, DataEntry] =
+    if (x.dataType == y.dataType && x.dataType == Boolean) formatResB[Boolean](Array(operator.op(x, y)), Boolean)
+    else Left(ContractInvalidOPCData)
+
+  def not(x: DataEntry): Either[ValidationError, DataEntry] =
+    x.dataType match {
+      case Boolean => formatResB[Boolean](Array(bNot(x)), Boolean)
+      case _ => Left(ContractUnsupportedOPC)
```

Comment:
`ContractInvalidOPCData`?

Reply:
if we suppose not(Int), not(Long) etc. will be valid someday like C language does in future, then it is `ContractUnsupportedOPC`. Otherwise it is `ContractInvalidOPCData`

Reply:
If we keep using ContractUnsupportedOPC here, we should separate the above boolBiOperation to three cases, right?

Reply:
or can not use ContractInvalidOPCData but ContractUnsupportedOPC and ContractDataTypeMismatch

Reply:
No, the function name is `boolBiOperation`, so the datatype which is not Boolean, is and will not be accepted. But `not` here is not related with any datatype, so it may accept other datatypes.

Reply:
Accept, clear now~